src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -327,7 +329,114 @@ roundRate rate rounding =
+  Party -> Party -> Optional Text -> Optional [Date] -> [(Decimal, Bool)] -> Update [TaggedClaim]
+calculateCompoundedFloatingPaymentClaimsFromSwapStream floatingRateCalculation s periodicSchedule
+  calculationSchedule paymentSchedule useAdjustedDatesForDcf issuerPaysLeg currency issuer
+  calendarDataAgency fxRateId fxFixingDates notionals = do
+    assertMsg "rate rounding not yet supported" $ isNone floatingRateCalculation.finalRateRounding
+    (rateFixingDates, rateFixingCalendar) <- getRateFixingsAndCalendars s (fromSome s.resetDates)
+      calculationSchedule issuer calendarDataAgency
+
+    case fxFixingDates of
+      Some fxDates ->
+        assertMsg "fxFixingDates must match rateFixingDates" $ fxDates == rateFixingDates
+      None -> assertMsg "No fx fixing dates provided" True
+    let
+      fixLeg = False
+      calculationNotionalFixings = zip3 calculationSchedule notionals rateFixingDates
+
+      -- calculate the notional and fixing dates for a payment period
+      calculateNotionalAndFixings p =
+        (notional, fxAdjRequired, calculationNotionalFixingsFiltered)
+          where
+            calculationNotionalFixingsFiltered = filter (\(c, nf, f) ->
+              c.unadjustedEndDate <= p.unadjustedEndDate &&
+              c.unadjustedStartDate >= p.unadjustedStartDate)
+              calculationNotionalFixings
+
+            -- Ensure that all notionals and fxAdjRequireds are the same for each paymentPeriod
+            notionals = map (\(_, (n, _), _) -> n) calculationNotionalFixingsFiltered
+            n = if length (dedup notionals) <= 1 then head notionals
+              else error "notional must be constant within a payment period"
+            fxAdjRequireds = map (\(_, (_, fxAdjRequired), _) -> fxAdjRequired)
+              calculationNotionalFixingsFiltered
+            fxAdjRequired = if length (dedup fxAdjRequireds) <= 1 then
+              head fxAdjRequireds
+              else error "fxAdjRequired must be constant within a payment period"
+
+            notionalConst = Const n
+            notional = case fxRateId of
+              None -> notionalConst
+              Some fxRateId -> if fxAdjRequired then notionalConst * fxRateObs else notionalConst
+                where fxRateObs = Observe fxRateId
+
+      -- calculate floating rate claim for a payment period
+      createClaim p =
+        scale notional $ compoundedRates
+        where
+          (notional, fxAdjRequired, calculationNotionalFixingsFiltered) =
+            calculateNotionalAndFixings p
+
+          floatingRateSpread = case floatingRateCalculation.spreadSchedule of
+            [] -> 0.0
+            [se] -> se.initialValue
+            _ -> error "Multiple SpreadSchedules not yet supported"
+          referenceRateId = floatingRateCalculation.floatingRateIndex
+          regularRate = Observe referenceRateId + Const floatingRateSpread
+
+          foldRates acc (c, (n, fxAdjRequired), f) =
```

Comment:
If possible, I would separate this into
- a generic claim utility function that compounds a generic expression at different points in time (we could move it to the common utils)
- the `compoundRates` function that uses that inserts the right formula into the above utility function

Reply:
Same here, I have added it to the compounding issue: https://github.com/digital-asset/daml-finance/issues/517